storage/storage_image.go

```diff
@@ -459,6 +475,118 @@ func (s *storageImageDestination) PutBlob(ctx context.Context, stream io.Reader,
 	}, nil
 }
 
+type seekableToReaderAt struct {
+	stream  types.ImageSourceSeekable
+	srcInfo types.BlobInfo
+	ctx     context.Context
+}
+
+func (s *seekableToReaderAt) ReadAt(p []byte, off int64) (n int, err error) {
+	chunk := types.ImageSourceChunk{
+		Offset: uint64(off),
+		Length: uint64(len(p)),
+	}
+	reader, _, err := s.stream.GetBlobAt(s.ctx, s.srcInfo, []types.ImageSourceChunk{chunk})
+	if err != nil {
+		return 0, err
+	}
+	defer reader.Close()
+
+	total := 0
+	for {
+		n, err := reader.Read(p[total:])
+		if err != nil && err != io.EOF {
+			return 0, err
+		}
+		total += n
+		if err == io.EOF || total == len(p) {
+			break
+		}
+	}
+	return total, nil
+}
+
+func (s *storageImageDestination) getLayersCache() (map[string][]compression.FileMetadata, map[string]string, error) {
+	s.putBlobMutex.Lock()
+	defer s.putBlobMutex.Unlock()
+
+	if s.LayersMetadata != nil {
+		return s.LayersMetadata, s.LayersTarget, nil
+	}
+
+	allLayers, err := s.imageRef.transport.store.Layers()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	layersMetadata := make(map[string][]compression.FileMetadata)
+	layersTarget := make(map[string]string)
+	for _, r := range allLayers {
+		if r.Metadata == "" {
+			continue
+		}
+		var metadataLayer []compression.FileMetadata
+		if err := json.Unmarshal([]byte(r.Metadata), &metadataLayer); err != nil {
+			continue
+		}
+		layersMetadata[r.ID] = metadataLayer
+		target, err := s.imageRef.transport.store.DifferTarget(r.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		layersTarget[r.ID] = target
+	}
+	s.LayersMetadata = layersMetadata
+	s.LayersTarget = layersTarget
+
+	return layersMetadata, layersTarget, nil
+}
+
+func (s *storageImageDestination) PutBlobPartial(ctx context.Context, stream types.ImageSourceSeekable, srcInfo types.BlobInfo, cache types.BlobInfoCache) (types.BlobInfo, error) {
+	blobDigest := srcInfo.Digest
+
+	readerAt := seekableToReaderAt{
+		stream:  stream,
+		srcInfo: srcInfo,
+		ctx:     ctx,
+	}
+
+	data, err := compression.ReadChunkedZstdManifest(&readerAt, srcInfo.Size)
```

Comment:
The `PutBlobPartial` interface looks generic. Just blindly hard-coding Zstd here seems surprising. (Maybe all it takes is a MIME type check.)